### PR TITLE
[EditableText] Fix TextField crashed with composing and maxLength set

### LINKED ISSUE #25511: Flutter will crash when use TextField to input Chinese with maxLength
### Environment
iOS 12.1

image
### Language
Chinese
### Sample Code
```dart
TextField(
   maxLength: 2,
)
```
Very easy to reappear，such as input “hh” when use Chinese input method，it will crash without any log in Android Studio

image

When I input the second "h", the value in the `editable_text.dart` is this

image

Then I debug step by step, it crash in `platform_messages.dart`

image

Finally I get some log from Mac Controller.app 

image

### LINKED ISSUE #35106: TextField set maxLength,  the chinese pinyin will be cutted
```

Flutter 1.6.3-pre.8 • channel master • https://github.com/flutter/flutter.git
Framework • revision be2918ff30 (5 weeks ago) • 2019-05-22 18:14:01 -0700
Engine • revision 8dc3a4cde2
Tools • Dart 2.3.2 (build 2.3.2-dev.0.0 e3edfd36b2)


```

```


[✓] Flutter (Channel master, v1.6.3-pre.8, on Mac OS X 10.14.3 18D42, locale
    zh-Hans-CN)
    • Flutter version 1.6.3-pre.8 at /Users/dashuaifu/dev/flutter
    • Framework revision be2918ff30 (5 weeks ago), 2019-05-22 18:14:01 -0700
    • Engine revision 8dc3a4cde2
    • Dart version 2.3.2 (build 2.3.2-dev.0.0 e3edfd36b2)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/dashuaifu/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_SDK_ROOT = /Users/dashuaifu/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • ios-deploy 1.9.4
    • CocoaPods version 1.6.1

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 34.0.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1248-b01)

[✓] Connected device (2 available)
    • nell’s iPhone • 00008020-001B24283E99002E            • ios • iOS 12.2
    • iPhone 8      • 2E6C5C1A-B83B-4BB4-8AE1-030C9385A034 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

• No issues found!
```

Jun-26-2019 15-51-04

### LINKED ISSUE #63753: [TextField] Exception when delete texts on iOS while the TextEditingValue is composing
## Steps to Reproduce

1. `TextField(maxLength: 5)` (Value can be any integer greater than zero).
2. Using a keyboard with composing, such as the Chinese keyboard.
3. Input until the characters length more than the max length, and keep composing status.
4. At this time the characters might flashing, if this appears, press delete button.
5. The app crashed.

Kapture 2020-08-14 at 16 53 45

## Environment

Platform: iOS
Flutter version: `1.0.0`~`master`

### PR DESCRIPTION
**Pull request by AlexV525**
## Description

This is a fix for `EditableText` and `LengthLimitingTextInputFormatter`, several behaviors updated.

- `_formatAndSetValue` now detects the composing changes and update normally.
- `maxLength` (`LengthLimitingTextInputFormatter`) 's constraint will be temporary lifted if the value is composing.
- No more composing if the value reached constraint.

Without this fix, apps will crash if they triggered any strange composing delete. It happens from flutter `1.0.0` until now.

| Before | After |
|--------|------|
| Kapture 2020-08-14 at 16 53 45 | Kapture 2020-08-14 at 16 55 19 |

## Related Issues

Resolves #63753
Resolves #35106
Resolves #25511

## Tests

I added the following tests:

- *Length formatter will not cause crash while the TextEditingValue is composing* in editable text test.
- *Validate form should return correct validation if the value is composing* in form test.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

### CODE REVIEW

**Comment by AlexV525** [comment]
@googlebot I signed it!

**Comment by cv692001** [comment]
Hey I am very new to contributing  , I have read documentations but I am still confused that how to proceed

**Comment by goderbauer** [comment]
Looks like this change is causing some tests to fail: https://cirrus-ci.com/task/4806069081866240. This can either mean you're breaking existing functionality or the tests need to be updated. Can you please take a look at those?

Also, in order to accept this PR you will have to write a test to ensure that we never regress and accidentally re-introduce the bug into our code base. Can you please add some tests?

/cc @justinmc

**Comment by AlexV525** [comment]
Yeah I just forgot to put this into WIP.

**Comment by justinmc** [comment]
@cv692001 I'm not sure if you're asking in relation to this PR, but if you're just asking about contributing in general, feel free to reach out to me via the contact info in my Github profile.

**Comment by AlexV525** [comment]
There's something wrong with my develop environment and I couldn't run tests normally. Scheduled tests update for 2~3 days. Sorry for that. 😢

**Comment by AlexV525** [comment]
@goderbauer @justinmc I have a question about using the tester to simulate the composing input.

I wrote something like this:
```dart
final EditableTextState state =
  tester.state<EditableTextState>(find.byType(EditableText));

state.updateEditingValue(const TextEditingValue(text: '111 1'));
expect(state.currentTextEditingValue.composing, TextRange.empty);

state.updateEditingValue(const TextEditingValue(text: '111 1', composing: TextRange(start: 2, end: 4)));
expect(state.currentTextEditingValue.composing, const TextRange(start: 2, end: 4));

state.updateEditingValue(const TextEditingValue(text: '111 11', composing: TextRange(start: 2, end: 5)));
expect(state.currentTextEditingValue.composing, const TextRange(start: 2, end: 5));
```

When I tried to add more text with composing like '111 11' (6 characters) it ends up nothing happened. How can I achieve the effect just like the real device's? (Like the screenshot below)
Kapture 2020-08-14 at 16 55 19

**Comment by AlexV525** [comment]
Forget about it. I forgot that I run this on stable, now it works on the new one.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2023,12 +2023,28 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     // as the pre-formatting value of the previous pass (repeat call).
     final bool textChanged = _value?.text != value?.text;
     final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
+    final bool isRepeatText = value?.text == _lastFormattedUnmodifiedTextEditingValue?.text;
 
-    if (textChanged && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
+    // There's no need to format when the value start to composing and repeat composing.
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2028]: I think this is more clear, but let me know if it doesn't seem accurate.

```suggestion
    // There's no need to format when starting to compose or when continuing
    // an existing composition.
```

**Review by justinmc — COMMENTED** [review]
Thanks for fixing this!  I confirmed that this fixes the problem for me locally and that the test fails without the fix.

Just one small change on a comment.

I had noticed this problem previously in https://github.com/flutter/flutter/issues/35106 and I imagined that something like this behavior would be the best solution.  I've added that issue to the description of this PR, because I think it fixes that too.

We should watch out for this causing any undesirable behavior in any existing Flutter apps.  For example, with this PR, it's now possible to get an invalid value from a TextField using `textFieldController.text`.  That may cause some Flutter apps to receive invalid values when they previously could not, if they're not checking the valid state of the form.

**Comment by justinmc** [comment]
I worry that this is a problem too:

Screen Shot 2020-08-17 at 10 49 45 AM

Maybe we shouldn't count composing characters as part of the character limit?  Would a native Chinese speaker expect the behavior in the screenshot, or should the character count be zero until the composing characters are committed?

**Comment by justinmc** [comment]
CC @LongCatIsLooong

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -2023,12 +2023,28 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    final bool isValueComposing = value != null && value.composing != const TextRange(start: -1, end: -1);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2029]: IIRC `TextRange.isValid` does the same:
```suggestion
    final bool isValueComposing = value?.composing?.isValid ?? false;
```

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -2023,12 +2023,28 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    final bool isLastUnmodifiedValueComposing = _lastFormattedUnmodifiedTextEditingValue != null &&
+      _lastFormattedUnmodifiedTextEditingValue.composing != const TextRange(start: -1, end: -1);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2031]: ditto the using `TextRange.isValid` comment.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -2023,12 +2023,28 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+
+    if (((!isRepeatText && textChanged) || (!isRepeatComposing && !isValueComposing)) &&
+        widget.inputFormatters != null &&
+        widget.inputFormatters.isNotEmpty) {
       // Only format when the text has changed and there are available formatters.
       // Pass through the formatter regardless of repeat status if the input value is
       // different than the stored value.
       for (final TextInputFormatter formatter in widget.inputFormatters) {
+        // Skip format update while the editing value is composing with fixed length.
+        if ((formatter is LengthLimitingTextInputFormatter) &&
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2042]: If the problem is with that specific formatter, can we apply the fix to the formatter instead?

**Review by LongCatIsLooong — COMMENTED** [review]
IMO we shouldn't specialize `LengthLimitingTextInputFormatter`. The fix should be applied to the implementation of that formatter if possible. 

@justinmc it makes sense to me for the length of the text to transiently surpass the limit if there's composing text. The user should be able to input the same thing using IME as they can with copy paste. But I think this is going to be an issue if the user submits the text without finalizing the composing text. 

Other input formatters (even third party ones) may have this issue too, @AlexV525 did you by any chance look into why it crashes?

**Comment by AlexV525** [comment]
> We should watch out for this causing any undesirable behavior in any existing Flutter apps. For example, with this PR, it's now possible to get an invalid value from a TextField using `textFieldController.text`. That may cause some Flutter apps to receive invalid values when they previously could not, if they're not checking the valid state of the form.

@justinmc I'll see if I can add a test to ensure it won't happened.

> IMO we shouldn't specialize `LengthLimitingTextInputFormatter`. The fix should be applied to the implementation of that formatter if possible.

That make sense.

> Other input formatters (even third party ones) may have this issue too, @AlexV525 did you by any chance look into why it crashes?

Basically when the composing editing value is updated by the formatter, the composing `TextRange`'s `end` is more than the length of the texts.

Let's go back to the crashing screenshot. Such as `12345` (with `5` composing `TextRange(start: 3, end: 4)`), then type `6`, the value was updated to `12346 ` (with `6 ` composing). Since then the composing might be `TextRange(start: 3, end: 5)` which is more than the actual length. At this time, call delete and the substring is failed, and crashed.

**Comment by AlexV525** [comment]
After a few hours of investigation, I found that it's hard to land this fix directly on the `LengthLimitingTextInputFormatter`. Here is a strange behavior I saw:
1. Input aaaa, with the `TextField` displayed `a a a`.
2. Input another 'a', now the value in `_formatAndSetValue` becomes `a`, which means it's been truncated or updated from the method `_handleTextInputInvocation`.
3. Continuing input multiple a, then delete when it's over-sized, crashed again.

If we only release the constraint inside the `LengthLimitingTextInputFormatter`, value won't be truncated when the composing is end and I see no events related.

Maybe it's something with keyboard behavior and the only thing we can do is what we currently have done.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 4 above)
Comment (AlexV525) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Done it, the fix is now land of the specific formatter.

**Comment by AlexV525** [comment]
Now the changes is both landed on `LengthLimitingTextInputFormatter` and the `_formatAndSetValue`, to keep composing status unified.

**Comment by AlexV525** [comment]
> We should watch out for this causing any undesirable behavior in any existing Flutter apps. For example, with this PR, it's now possible to get an invalid value from a TextField using `textFieldController.text`. That may cause some Flutter apps to receive invalid values when they previously could not, if they're not checking the valid state of the form.

The last thing is about this concern. For me, technically, it should be the current value whether it's composing or not. I added a test about form validation when the value composing. Although it can return `false` if the composing value cannot pass the validator, we definitely should keep our eyes on related behaviors.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 5]
```diff
@@ -370,6 +370,10 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
     TextEditingValue oldValue, // unused.
     TextEditingValue newValue,
   ) {
+    // Always return new value when it's composing.
+    if (newValue.isComposingRangeValid) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_formatter.dart:374]: Using `isComposingRangeValid` is a bit too strong. It's mainly used for out of range check. To check whether `newValue` contains composing text, I think `newValue.composing.isValid` is preferred. Yeah I'm sorry the name is a bit confusing.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:374]: Also applies to the changes below.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:374]: I think the behavior change is worth documenting. We should probably add a "#Multistage text input" section or "#Composing Text" section somewhere in this class for visibility, and document the fact that the [maxLength] constraint is going to be temporarily lifted when the user starts composing, and when the composing finishes, the text may get truncated so it does not exceed [maxLength].

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:374]: Also, ideally we shouldn't allow the user to start composing when they've already reached the length limit.

**Review by LongCatIsLooong — COMMENTED** [review]
LGTM in general. Thank you for contributing! 

I check the current close connection logic, the composing range will be cleared when the current input connection closes. So when the user switches to a different text field, with the changes in this PR the `LengthLimitingTextInputFormatter` will kick in and truncate the text if needed. This is great. :+1: .

File: packages/flutter/lib/src/services/text_formatter.dart (hunk 5 above)
Comment (AlexV525) [reply packages/flutter/lib/src/services/text_formatter.dart:374]: > Also, ideally we shouldn't allow the user to start composing when they've already reached the length limit.

Strongly agreed with this one! I'll add it.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 6]
```diff
@@ -2024,7 +2024,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
     final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
 
-    if (textChanged && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
+    // There's no need to format when starting to compose or when continuing
+    // an existing composition.
+    final bool isValueComposing = value?.isComposingRangeValid ?? false;
+    final bool isLastUnmodifiedValueComposing = _lastFormattedUnmodifiedTextEditingValue?.isComposingRangeValid ?? false;
+    final bool isRepeatComposing = isValueComposing == isLastUnmodifiedValueComposing;
+
+    if ((textChanged || (!isRepeatComposing && !isValueComposing)) &&
```
[the remark this answers is not in the document]
Comment (AlexV525) [reply packages/flutter/lib/src/widgets/editable_text.dart:2033]: That is somehow same as above fields "isValue*". 🤣

[the remark this answers is not in the document]
Comment (AlexV525) [reply packages/flutter/lib/src/widgets/editable_text.dart:2033]: > `!isComposing && isPreviouslyComposing`

Conditions are really a science.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 7]
```diff
@@ -367,9 +376,19 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
 
   @override
   TextEditingValue formatEditUpdate(
-    TextEditingValue oldValue, // unused.
```
Comment (AlexV525) [inline packages/flutter/lib/src/services/text_formatter.dart:370]: I've removed this comment, it seems useless.

**Comment by AlexV525** [comment]
I've updated the description for behavior changes.

**Review by justinmc — APPROVED** [review]
LGTM 👍   Just a quick grammar fix in a comment below.

It looks like Windows_tool_tests is failing, but it seems unrelated. I tried rerunning it, but if it doesn't pass, it might be necessary to update the branch with master.

**Comment by AlexV525** [comment]
(knock knock...)

**Review by justinmc — APPROVED** [review]
LGTM 👍 

This should be merged when the build is green.  Thanks for all the work on this!

**Pull request merged**